craft_application/secrets.py

```diff
@@ -0,0 +1,118 @@
+# This file is part of craft_application.
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Handling of build-time secrets."""
+from __future__ import annotations
+
+import re
+import subprocess
+from typing import Any
+
+from craft_application import errors
+
+SECRET_REGEX = re.compile(r"\$\(HOST_SECRET:(?P<command>.*)\)")
+
+
+def render_secrets(yaml_data: dict[str, Any]) -> set[str]:
+    """Render/expand the build secrets in a project's yaml data (in-place).
+
+    This function will process directives of the form $(HOST_SECRET:<cmd>) in string
+    values in ``yaml_data``. For each such directive, the <cmd> part is executed (with
+    bash) and the resulting output replaces the whole directive. The returned set
+    contains the outputs of all HOST_SECRET processing (for masking with craft-cli).
+
+    Note that only a few fields are currently supported:
+
+    - "source" and "build-environment" for parts.
+
+    Using HOST_SECRET directives in any other field is an error.
+    """
+    command_cache: dict[str, str] = {}
+
+    # Process the fields where we allow build secrets
+    parts = yaml_data.get("parts", {})
+    for part in parts.values():
+        _render_part_secrets(part, command_cache)
+
+    # Now loop over all the data to check for build secrets in disallowed fields
+    _check_for_secrets(yaml_data)
+
+    return set(command_cache.values())
+
+
+def _render_part_secrets(
```

Comment:
fwiw, we have a lot of work on the extensions side to make seamless operations against scalars and iterables which can be brought in here once we +1 the keys we support